File: Graph/php/shortestPath.php
<?php

declare(strict_types=1);

require_once __DIR__ . '/GraphUtils.php';

function shortestPath(array $edges, string $nodeA, string $nodeB): int
{
    $graph = buildUndirectedGraph($edges);
    $queue = [[$nodeA, 0]];
    $visited = [$nodeA => true];

    while ($queue) {
        [$node, $distance] = array_shift($queue);
        if ($node === $nodeB) {
            return $distance;
        }

        foreach ($graph[$node] ?? [] as $neighbor) {
            if (isset($visited[$neighbor])) {
                continue;
            }
            $visited[$neighbor] = true;
            $queue[] = [$neighbor, $distance + 1];
        }
    }

    return -1;
}

File: Graph/java/ShortestPath.java
import java.util.ArrayDeque;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Queue;
import java.util.Set;

public class ShortestPath {
  public static int shortest(List<String[]> edges, String nodeA, String nodeB) {
    Map<String, List<String>> graph = GraphUtils.buildUndirectedGraph(edges);
    Queue<NodeDistance> queue = new ArrayDeque<>();
    Set<String> visited = new HashSet<>();
    queue.add(new NodeDistance(nodeA, 0));
    visited.add(nodeA);

    while (!queue.isEmpty()) {
      NodeDistance nd = queue.poll();
      if (nd.node().equals(nodeB)) {
        return nd.distance();
      }

      for (String neighbor : graph.getOrDefault(nd.node(), List.of())) {
        if (visited.add(neighbor)) {
          queue.add(new NodeDistance(neighbor, nd.distance() + 1));
        }
      }
    }

    return -1;
  }

  private record NodeDistance(String node, int distance) {}
}
